src/pwm.rs

```diff
@@ -0,0 +1,186 @@
+            Err(()) => Err(()),
+        }
+    }
+
+    /// Creates a PWM device from an already initialized c type.
+    ///
+    /// **Note** that the `CHANNELS` Variable is only used to cache the duty values for each of the `CHANNELS` number of
+    /// pwm channels. This is used by [`embedded_hal::Pwm::get_duty`]. If this is not needed setting `CHANNELS` to `0` saves memory.
+    ///
+    /// ## Important:
+    /// It is **important** to make sure that the provided **device is already initialized** by using [`riot_sys::pwm_init`](https://rustdoc.etonomy.org/riot_sys/fn.pwm_init.html).
+    /// Using the returned device otherwise results in **undefined behavior**!
+    ///
+    /// Also note, that the given frequency is only to be used by [`embedded_hal::Pwm::get_period`] or [`PWMDevice::get_frequency`].
+    /// Just setting this to `x` will only result in those two functions returning `x` but will have no other impact, if
+    /// for example the given [`pwm_t`] is initialized by a board and the actually set frequency is unknown.
+    ///
+    pub unsafe fn new_without_init(dev: pwm_t, frequency: Hz) -> Self {
+        PWMDevice {
+            dev,
+            channel_duty_values: [0; CHANNELS],
+            frequency,
+        }
+    }
+
+    /// Initializes the [`PWMDevice`] with the given frequency in Hz and resolution of a period/duty cycle.
+    ///
+    /// If the given frequency and resolution values can not be acchieved on the current device, the
+    /// resolution will be kept the same while the frequency will be lowered until the device can handle the combination.
+    /// The resulting frequency is written into the [`PWMDevice`].
+    ///
+    /// Uses a [`Result`] in anticipation of a more usefull error handling in the future.
+    ///
+    /// Returns if the initialization was a success
+    fn init(&mut self, mode: PWMMode, frequency: Hz, resolution: u16) -> Result<(), ()> {
+        let err = unsafe { riot_sys::pwm_init(self.dev, mode.to_c(), frequency, resolution) };
+
+        match err {
+            0 => Err(()),
+            freq => {
+                // Set frequency in the device
+                self.frequency = freq;
+                Ok(())
+            }
+        }
+    }
+
+    /// Returns the number of available channels for this device
+    pub fn channels(&self) -> u8 {
+        unsafe { riot_sys::pwm_channels(self.dev).min(CHANNELS as u8) }
+    }
+
+    /// Stops PWM generation on this device
+    pub fn power_off(&mut self) {
+        unsafe { riot_sys::pwm_poweroff(self.dev) }
+    }
+
+    /// Resumes PWM generation after power_off on this device
+    pub fn power_on(&mut self) {
+        unsafe {
+            riot_sys::pwm_poweron(self.dev);
+        }
+    }
+
+    /// Sets the duty-cycle for the given channel
+    ///
+    /// value: `0: 0%, u16::MAX: 100%` duty_cycle
+    pub fn set(&mut self, channel: u8, value: u16) {
```

Comment:
On other peripherals the methods that can be used through embedded_hal are note even in the main impl, at least not pub'ly. Is there a particular reason why this should be pub?

Reply:
I first had in mind that if one would not use `embedded_ hal` there would be a complete separated API, but maybe it's just annoying to have this methods basically duplicated. So only the power methods and `channels` should remain in the pub api, since they can not be accessed from `embedded_hal` at the moment.

Reply:
Removed the `pub` and deleted `get_frequency` since that is not used anymore. (See https://github.com/RIOT-OS/rust-riot-wrappers/pull/38/commits/8d770fd642f3b92c3b94a321de643fe8929d30ff)

Reply:
Thanks.

I'm not 100% sure we're on the correct path here, but it has worked well so far with the other peripherals.